lib/pstore.rb

```diff
@@ -10,87 +10,318 @@
+# - {Transactions}[rdoc-ref:PStore@Transactions]: each transaction is a collection
+#   of prospective changes to the store;
+#   a transaction is defined in the block given with a call
+#   to PStore#transaction.
+#
+# == About the Examples
+#
+# Examples on this page need a store that has known properties.
+# Therefore examples may assume the existence of this helper method:
+#
+#   require 'pstore'
+#   require 'tempfile'
+#   # Yield a pristine store for use in examples.
+#   def example_store
+#     # Create the store in a temporary file.
+#     path = Tempfile.create
+#     store = PStore.new(path)
+#     # Populate the store.
+#     store.transaction do
+#       store[:foo] = 0
+#       store[:bar] = 1
+#       store[:baz] = 2
+#     end
+#     yield store
+#   end
+#
+# This means that example code can look like this:
+#
+#   example_store do |store|
+#     # Example code using store goes here.
+#   end
+#
+# == The Store
+#
+# The contents of the store are maintained in a file whose path is specified
+# when the store is created (see PStore.new).
+# The objects are stored and retrieved using
+# module Marshal, which means that certain objects cannot be added to the store;
+# see {Marshal::dump}[https://docs.ruby-lang.org/en/master/Marshal.html#method-c-dump].
+#
+# == Roots
+#
+# A store may have any number of entries, called _roots_.
+# Each root has a key and a value, just as in a hash:
+#
+# - Key: as in a hash, the key can be (almost) any object;
+#   see {Hash Keys}[https://docs.ruby-lang.org/en/master/Hash.html#class-Hash-label-Hash+Keys].
+#   You may find it convenient to keep it simple by using only
+#   symbols or strings as keys.
+# - Value: the value may be any object that can be marshalled by \Marshal
+#   (see {Marshal::dump}[https://docs.ruby-lang.org/en/master/Marshal.html#method-c-dump])
+#   and in fact may be a collection
+#   (e.g., an array, a hash, a set, a range, etc).
+#   That collection may in turn contain nested objects,
+#   including collections, to any depth;
+#   those objects must also be \Marshal-able.
+#   See {Deep Root Values}[rdoc-ref:PStore@Deep+Root+Values].
+#
+# == Transactions
+#
+# === The Transaction Block
+#
+# The block given with a call to method #transaction#
+# contains a _transaction_,
+# which consists of calls to \PStore methods that
+# read from or write to the store
+# (that is, all \PStore methods except #transaction itself,
+# #path, and Pstore.new):
+#
+#   example_store do |store|
+#     store.transaction do
+#       store.roots # => [:foo, :bar, :baz]
+#       store[:bat] = 3
+#       store.roots # => [:foo, :bar, :baz, :bat]
+#     end
+#   end
+#
+# Execution of the transaction is deferred until the block exits,
+# and is executed _atomically_ (all-or-nothing):
+# either all transaction calls are executed, or none are.
+# This maintains the integrity of the store.
+#
+# Other code in the block (including even calls to #path and PStore.new)
+# is executed as it is read.
```

Comment:
I'm confused by this sentence. I'm not sure what it means.

Reply:
Maybe "... executed immediately, not deferred."

Reply:
The point I was trying for is not that unrelated code can be in the block, but that (unlike the transaction calls) that code's execution is not deferred.

Reply:
Yeah maybe it would be useful to mention that it's executed immediately.

Reply:
@BurdetteLamar Did you miss this?

Reply:
Modified.